### PR TITLE
CAP-28 Search Meals pages

### PR DESCRIPTION
**Pull request by snaanortey**
Here is what I did: 

Short video of UI - https://user-images.githubusercontent.com/85527341/209823455-92cf2b15-e7bb-4a17-a74d-75162075f3b3.mov

`SearchMeal` page - Allows the user to type a meal. Populate suggestions as the user types 3 or more characters.
`ShoppingList/[mealID]` page - Provides a list of ingredients for the meal the user selects on the `SearchMeal` page

### CODE REVIEW

File: client2/pages/searchMeal/index.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      const { data } = await axios.get(
+        `http://localhost:8000/meals/search?query=${mealName}`
+      );
+
+      setSuggestedMeal(data);
+    }
+
+    if (mealName.length <= 2) {
+      setSuggestedMeal([]);
+    }
```
Comment (ehimah) [inline client2/pages/searchMeal/index.tsx:29]: what does this do?

File: client2/pages/searchMeal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  const getSuggestedMeal: React.ChangeEventHandler<HTMLInputElement> = async (
+    e
+  ) => {
+    const mealName = e.target.value;
+
+    if (mealName.length >= 3) {
+      const { data } = await axios.get(
+        `http://localhost:8000/meals/search?query=${mealName}`
+      );
```
Comment (ehimah) [inline client2/pages/searchMeal/index.tsx:22]: please bring in the backend class. all calls to your backend should be done using that `BckendClient` class

File: client2/pages/shoppingList/[recipeId].tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  title: string;
+}
+
+const ShoppingList: NextPageWithLayout = () => {
+  const router = useRouter();
+  const { recipeId } = router.query;
+  const [shoppingList, setShoppingList] = useState<string[]>([]);
+
+  axios
+    .get(`http://localhost:8000/ingredients/${recipeId}`)
```
Comment (ehimah) [inline client2/pages/shoppingList/[recipeId].tsx:23]: same here

File: client2/pages/shoppingList/[recipeId].tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    .then((response) => {
+      setShoppingList(response.data.ingredients);
+    })
+    .catch((err) => {
+      console.log(err);
+    });
+
+  if (!shoppingList) {
+    <p>Shopping list is loading...</p>;
+  }
```
Comment (ehimah) [inline client2/pages/shoppingList/[recipeId].tsx:33]: We need to invest in some aesthetics and best practices for loading spinners using React Suspense and/or lazy loading across the entire app.

See references https://beta.nextjs.org/docs/data-fetching/streaming-and-suspense
and https://learnjsx.com/category/4/posts/nextjs-suspense

Could you create a JIRA ticket for it and link it here?

File: client2/pages/searchMeal/index.tsx (hunk 1 above)
Comment (ehimah) [inline client2/pages/searchMeal/index.tsx:29]: should this be an else?

**Review by ehimah — CHANGES_REQUESTED** [review]
Some comments and we're good.

File: client2/pages/searchMeal/index.tsx (hunk 1 above)
Comment (snaanortey) [reply client2/pages/searchMeal/index.tsx:29]: fixed this

File: client2/pages/searchMeal/index.tsx (hunk 2 above)
Comment (snaanortey) [reply client2/pages/searchMeal/index.tsx:22]: Done this. Included for other APIs.

File: client2/pages/shoppingList/[recipeId].tsx (hunk 4 above)
Comment (snaanortey) [reply client2/pages/shoppingList/[recipeId].tsx:33]: Created the ticket [here](https://naanorkor.atlassian.net/browse/CAP-34?atlOrigin=eyJpIjoiYjZmZDUxNTg3NmVjNGNmZTg5NzU2M2UyNWE4YTNhOTAiLCJwIjoiaiJ9)

File: client2/pages/searchMeal/index.tsx (hunk 1 above)
Comment (snaanortey) [reply client2/pages/searchMeal/index.tsx:29]: Updated

**Comment by snaanortey** [comment]
Please re-review

File: client2/pages/searchMeal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    <div className="m-4">
+      <input
+        placeholder="type a meal, get your shopping list"
+        onChange={getSuggestedMeal}
+        className={`bg-gray-200 w-full p-2 border-teal-500 border-spacing-4 border-2 rounded-xl dark:md:hover:placeholder-fuchsia-600
+              hover:bg-slate-400 hover:cursor-pointer`}
+      ></input>
+      <div>
+        {suggestedMeal.map((meal, i) => (
+          <Link key={i} href={`/shoppingList/${meal.id}`}>
```
Comment (ehimah) [inline client2/pages/searchMeal/index.tsx:37]:
```suggestion
        {suggestedMeal.map(({id, title}, i) => (
          <Link key={i} href={`/shoppingList/${id}`}>
```

File: client2/pages/searchMeal/index.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    e
+  ) => {
+    const mealName = e.target.value;
+
+    if (mealName.length >= 3) {
+      const data = await backendService.searchMealName(mealName);
+      setSuggestedMeal(data);
+    } else {
+      setSuggestedMeal([]);
+    }
```
Comment (ehimah) [inline client2/pages/searchMeal/index.tsx:24]:
```suggestion

    if (mealName.length < 3) {
      setSuggestedMeal([]);
      return;
    } 
    const data = await backendService.searchMealName(mealName);
      setSuggestedMeal(data);
```

File: client2/services/backend.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import axios, { AxiosStatic } from 'axios';
+import { MealSearchResult, RecipePerMealSelected } from '../types';
+
+class BackendService {
+  private httpClient: AxiosStatic;
+  private baseUrl: string;
+  private bearerToken: string | null;
+
+  constructor() {
+    this.baseUrl = 'http://localhost:8000';
```
Comment (ehimah) [inline client2/services/backend.ts:10]: We should make this configurable via constructor arguments like
```suggestion
  constructor(apiBaseUrl: string) {
    this.baseUrl = apiBaseUrl ;
```



this way we can later instantiate the class as 

```ts
const service = new BackendService('http://localhost:8000');

// or even using an environment variable
const service = new BackendService(process.env.BACKEND_API_URL);
```

File: client2/services/backend.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      const { data } = await this.httpClient.get(url);
+      return data;
+    } catch (err) {
+      console.log(err);
+      throw err;
+    }
+  };
+}
+
+export const backendService = new BackendService();
```
Comment (ehimah) [inline client2/services/backend.ts:51]: this should change and the api url should be passed in

**Review by ehimah — COMMENTED** [review]
a few comments to address

File: client2/pages/searchMeal/index.tsx (hunk 5 above)
Comment (snaanortey) [reply client2/pages/searchMeal/index.tsx:37]: 👍🏾 updated this

File: client2/pages/searchMeal/index.tsx (hunk 6 above)
Comment (snaanortey) [reply client2/pages/searchMeal/index.tsx:24]: 👍🏾 done

File: client2/services/backend.ts (hunk 7 above)
Comment (snaanortey) [reply client2/services/backend.ts:10]: updated

**Comment by snaanortey** [comment]
All comments addressed

**Review by ehimah — APPROVED** [review]
👌🏽

**Pull request merged**